yb-voyager/cmd/importData.go

```diff
@@ -820,44 +788,6 @@ func getTablesToImport() ([]string, []string, []string, error) {
 	return doneTables, interruptedTables, remainingTables, nil
 }
 
-func doImport(taskQueue chan *SplitFileImportTask, parallelism int, connPool *tgtdb.ConnectionPool) {
-	if Done.IsSet() { //if import is already done, return
-		log.Infof("Done is already set.")
-		return
-	}
-	parallelImportCount := int64(0)
-
-	importProgressContainer = ProgressContainer{
-		container: mpb.New(),
-	}
-	if !disablePb {
-		go importDataStatus()
-	}
-
-	for Done.IsNotSet() {
-		select {
-		case t := <-taskQueue:
-			// fmt.Printf("Got taskfile = %s putting on parallel channel\n", t.SplitFilePath)
-			// parallelImportChannel <- t
-			for parallelImportCount >= int64(parallelism) {
-				time.Sleep(time.Second * 2)
-			}
-			atomic.AddInt64(&parallelImportCount, 1)
-			go doImportInParallel(t, connPool, &parallelImportCount)
-		default:
-			// fmt.Printf("No file sleeping for 2 seconds\n")
-			time.Sleep(2 * time.Second)
-		}
-	}
-
-	importProgressContainer.container.Wait()
-}
-
-func doImportInParallel(t *SplitFileImportTask, connPool *tgtdb.ConnectionPool, parallelImportCount *int64) {
-	doOneImport(t, connPool)
-	atomic.AddInt64(parallelImportCount, -1)
-}
-
 func doOneImport(task *SplitFileImportTask, connPool *tgtdb.ConnectionPool) {
 	log.Infof("Importing %q", task.SplitFilePath)
 	//this is done to signal start progress bar for this table
```

Comment:
This is refactored as below in the `new-state` branch:
```
	copyCommand := getCopyCommand(batch.TableName)
	// copyCommand is empty when there are no rows for that table
	if copyCommand == "" {
		err = batch.MarkDone()
		if err != nil {
			utils.ErrExit("marking batch %q as done: %s", batch.FilePath, err)
		}
		return
	}
```
I don't understand the corner case you're mentioning. And also, I don't understand the comment in the code as well 🤔 -- when the control is in `doOneImport()` a batch is already prepared; how can the table does not have any rows.

For now, I will keep the code as it is.

Reply:
No, table can have rows but i am talking unable to find out the COPY commands in case of pg_dump/ora2pg.
Since we use regexp in those to look into a file `toc.txt` and `_data.sql` respectively, and if somehow our regexp fails, maybe due to some change in syntax of pg_dump version being used. Then our code won't be able to find copy command even though there are rows and it will mark and report it as done to the user.
This section of code from `extractCopyStmtForTable` function
```
var copyCommandRegex *regexp.Regexp
	if sourceDBType == "postgresql" {
		// find the line from toc.txt file
		fileToSearchIn = exportDir + "/data/toc.txt"

		conn := newTargetConn()
		defer conn.Close(context.Background())

		parentTable := getParentTable(table, conn)
		tableInCopyRegex := table
		if parentTable != "" { // needed to find COPY command for table partitions
			tableInCopyRegex = parentTable
		}
		if len(strings.Split(tableInCopyRegex, ".")) == 1 {
			// happens only when table is in public schema, use public schema with table name for regexp
			copyCommandRegex = regexp.MustCompile(fmt.Sprintf(`(?i)COPY public.%s[\s]+\(.*\) FROM STDIN`, tableInCopyRegex))
		} else {
			copyCommandRegex = regexp.MustCompile(fmt.Sprintf(`(?i)COPY %s[\s]+\(.*\) FROM STDIN`, tableInCopyRegex))
		}
		log.Infof("parentTable=%s for table=%s", parentTable, table)
	} else if sourceDBType == "oracle" || sourceDBType == "mysql" {
		// For oracle, there is only unique COPY per datafile
		// In case of table partitions, parent table name is used in COPY
		copyCommandRegex = regexp.MustCompile(`(?i)COPY .*[\s]+\(.*\) FROM STDIN`)
	}
```

Reply:
I know this is an old code and not part of this refactoring, if you agree with this corner case, we can do this change in a separte PR